### PR TITLE
refactor(catalog): refactor system catalog using proc macro

### LINKED ISSUE #13433: Proposal: `#[Catalog]` derive proc macro
I'd like to propose a `[Catalog]` derive proc macro to simplify the implementation of our system catalog.

There are three current pain points:

1. We have to construct the schema manually.
2. We must ensure the returned `OwnedRow` is consistent with the schema.
3. We have to register it manually.

## Type of row

```rust
#[derive(Catalog)]
#[catalog(reader, rw_catalog.rw_schemas)]
pub struct RwSchemasRow {
    id: i32,
    name: String,
    owner: i32,
    acl: String,
}
```

This will generate a static method on it:

```rust
impl RwSchemaRow {
    pub fn columns() -> Vec<SystemCatalogColumnsDef<'static>> { ... }
}
```

## `[catalog]` macro attribute

The `catalog` macro will do three things:

1. Generate a `BuiltinView` or `BuiltinTable` based on given information.
2. Register it to the global `SYS_CATALOGS`.

## Reader

We can implement the reader now with type safety.

```rust
#[derive(Catalog)]
#[catalog(reader, rw_catalog.rw_schemas)]
pub struct RwSchemasRow {
    id: i32,
    name: String,
    owner: i32,
    acl: String,
}

impl RwSchemasRow {
/// A read method is necessary for a Reader.
pub fn read(reader: &SysCatalogReaderImpl) -> Result<Vec<RwSchemasRow>> {
    let catalog = reader.catalog_reader.read_guard();
    let schemas = catalog.iter_schemas(&reader.auth_context.database)?;
    let user_reader = catalog.user_info_reader.read_guard();
    let users = user_reader.get_all_users();
    let username_map = user_reader.get_user_name_map();

    Ok(schemas
        .map(|schema| {
            RwSchemasRow {
                id: schema.id() as i32,
                name: schema.name().into(),
                owner: schema.owner() as i32,
                acl: get_acl_items(&Object::SchemaId(schema.id()), false, &users, username_map).into(),
            }
        })
        .collect_vec())
}

}
```

## View

Only the column types need to be provided in the function signature.

```rust
#[derive(Catalog)]
#[catalog(view, pg_catalog.pg_indexes,
  r#"SELECT s.name AS schemaname, \
        t.name AS tablename, \
        i.name AS indexname, \
        NULL AS tablespace, \
        i.definition AS indexdef \
    FROM rw_catalog.rw_indexes i \
    JOIN rw_catalog.rw_tables t ON i.primary_table_id = t.id \
    JOIN rw_catalog.rw_schemas s ON i.schema_id = s.id"#
)]
pub struct PgIndexesRow {
   // ...
}
```

## Dummy

Similar to View:

```rust
#[derive(Catalog)]
#[catalog(dummy, pg_catalog.pg_aggregate)]
pub struct PgAggregateRow {
   // ...
}
```

### PR DESCRIPTION
**Pull request by wangrunji0408**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

resolve #13433 following #14934

This PR refactors system catalog using the proc macro proposed in #13433, but in a slightly different way.

Instead of `#[derive(Catalog)]`, this PR introduces `#[system_catalog]` macro (proposed by https://github.com/risingwavelabs/risingwave/issues/13433#issuecomment-1811847857), which can be put on a function to define system table, or put on a struct to define system view.

```rust
// define a dummy system view
#[system_catalog(view, "pg_catalog.pg_am")]
#[derive(Fields)]
struct PgAmColumn { ... }

// define a system view with SQL query
#[system_catalog(view, "information_schema.tables", "SELECT ...")]
#[derive(Fields)]
struct Table { ... }

// define a system table
#[system_catalog(table, "rw_catalog.rw_types")]
fn read_rw_types(reader: &SysCatalogReaderImpl) -> Result<Vec<RwType>> { ... }
```

Similar to `#[function]`, these tables & views will be automatically registered into the global `SYS_CATALOGS`.

Table/view ids are generated in lexicographic order of their name. Thus ids are unstable across different versions. Not sure if this is safe. Or we should assign static ids to them?

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

**Comment by BugenZhao** [comment]
Wow, this is a huge project. I'm curious how did you do such refactor efficiently? 😄

**Comment by wangrunji0408** [comment]
> Wow, this is a huge project. I'm curious how did you do such refactor efficiently? 😄

Thanks to Github Copilot. 😄

WX20240202-142538@2x

File: src/batch/src/executor/sys_row_seq_scan.rs [hunk 1]
```diff
@@ -107,26 +105,13 @@ impl Executor for SysRowSeqScanExecutor {
 impl SysRowSeqScanExecutor {
     #[try_stream(boxed, ok = DataChunk, error = BatchError)]
     async fn do_executor(self: Box<Self>) {
-        let rows = self
+        let chunk = self
             .sys_catalog_reader
             .read_table(&self.table_id)
             .await
             .map_err(BatchError::SystemTable)?;
-        let filtered_rows = rows
```
Comment (yezizp2012) [inline src/batch/src/executor/sys_row_seq_scan.rs:115]: Deleted by mistake? 👀

**Comment by wangrunji0408** [comment]
Holy

image

**Review by yezizp2012 — APPROVED** [review]
LSTM!! rubber stamp for all writes on system catalog files.

**Review by TennyZhuang — APPROVED** [review]
Approved as lock team member.

**Pull request merged**